File: flink-demo/src/main/scala/com/darcy/flink/SocketTextStreamWordCount.scala
package com.darcy.flink

import org.apache.flink.api.common.functions.FlatMapFunction
import org.apache.flink.api.java.tuple.Tuple2
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment
import org.apache.flink.util.Collector;

final class LineSplitter extends FlatMapFunction[String, Tuple2[String, Integer]] {
  override def flatMap(s: String, collector: Collector[Tuple2[String, Integer]]): Unit = {
    val tokens = s.toLowerCase.split("\\W+")
    for (token <- tokens) {
      if (token.length > 0) collector.collect(new Tuple2[String, Integer](token, 1))
    }
  }
}

object SocketTextStreamWordCount {

  def main(args: Array[String]): Unit = {
    if (args.length != 2) {
      println("USAGE:\nSocketTextStreamWordCount <hostname> <port>")
      return
    }

    val hostName = args(0)
    val port = args(1).toInt

    val env = StreamExecutionEnvironment.getExecutionEnvironment
    val stream = env.socketTextStream(hostName, port)

    val sum = stream.flatMap(new LineSplitter())
      .keyBy(0)
      .sum(1)

    sum.print()
    env.execute("Scala WordCount From SocketTextStream Example.")

  }
}

File: flink-demo/src/main/java/com/darcy/splitter/SocketTextStreamWordCount.java
package com.darcy.splitter;

import org.apache.flink.api.common.functions.FlatMapFunction;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.streaming.api.datastream.DataStreamSource;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.util.Collector;

/**
 * 运行命令:
 * 1. $ tail -f flink-darcy-taskexecutor-0-darcy.local.out
 * 2. $ nc -l 172.30.11.248  9100
 * hello
 * hello
 * ...
 * 3. $ flink run -c com.darcy.splitter.SocketTextStreamWordCount flink-demo-1.0-SNAPSHOT.jar 172.30.11.248 9100
 * Starting execution of program
 *
 * 先执行 nc -l 172.30.11.248  9100, 再flink run， 此时再在nc中发送消息;
 */
public class SocketTextStreamWordCount {
    public static void main(String[] args) throws Exception {
        //参数检查
        if (args.length != 2) {
            System.err.println("USAGE:\nSocketTextStreamWordCount <hostname> <port>");
            return;
        }

        String hostname = args[0];
        Integer port = Integer.parseInt(args[1]);


        // set up the streaming execution environment
        final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();

        //获取数据
        DataStreamSource<String> stream = env.socketTextStream(hostname, port);

        //计数
        SingleOutputStreamOperator<Tuple2<String, Integer>> sum = stream.flatMap(new LineSplitter())
                .keyBy(0)
                .sum(1);

        sum.print();

        env.execute("Java WordCount from SocketTextStream Example");
    }

    public static final class LineSplitter implements FlatMapFunction<String, Tuple2<String, Integer>> {
        @Override
        public void flatMap(String s, Collector<Tuple2<String, Integer>> collector) {
            String[] tokens = s.toLowerCase().split("\\W+");

            for (String token: tokens) {
                if (token.length() > 0) {
                    collector.collect(new Tuple2<>(token, 1));
                }
            }
            System.out.println("flatMap value:" + s);
        }
    }
}
